appliance/mariadb/cmd/flynn-mariadb-api/main.go

```diff
@@ -119,9 +134,148 @@ func (a *API) dropDatabase(ctx context.Context, w http.ResponseWriter, req *http
 }
 
 func (a *API) ping(ctx context.Context, w http.ResponseWriter, req *http.Request) {
-	if _, err := a.db.Exec("SELECT 1"); err != nil {
+	db, err := a.connect()
+	if err != nil {
+		httphelper.Error(w, err)
+		return
+	}
+	defer db.Close()
+
+	if _, err := db.Exec("SELECT 1"); err != nil {
 		httphelper.Error(w, err)
 		return
 	}
 	w.WriteHeader(200)
 }
+
+func (a *API) connect() (*sql.DB, error) {
+	dsn := &mariadb.DSN{
+		Host:     serviceHost + ":3306",
+		User:     "flynn",
+		Password: os.Getenv("MYSQL_PWD"),
+		Database: "mysql",
+	}
+	return sql.Open("mysql", dsn.String())
+}
+
+func (a *API) scaleUp() error {
+	app := "mariadb"
+	logger := a.logger().New("fn", "scaleUp")
+
+	// Connect to controller.
+	logger.Info("connecting to controller")
+	client, err := controller.NewClient("", os.Getenv("CONTROLLER_KEY"))
+	if err != nil {
+		logger.Error("controller client error", "err", err)
+		return err
+	}
+
+	// Retrieve mariadb release.
+	logger.Info("retrieving app release", "app", app)
+	release, err := client.GetAppRelease(app)
+	if err == controller.ErrNotFound {
+		logger.Error("release not found", "app", app)
+		return errors.New("mariadb release not found")
+	} else if err != nil {
+		logger.Error("get release error", "app", app, "err", err)
+		return err
+	}
+
+	// Retrieve current formation.
+	logger.Info("retrieving formation", "app", app, "release_id", release.ID)
+	formation, err := client.GetFormation(app, release.ID)
+	if err == controller.ErrNotFound {
+		logger.Error("formation not found", "app", app, "release_id", release.ID)
+		return errors.New("mariadb formation not found")
+	} else if err != nil {
+		logger.Error("formation error", "app", app, "release_id", release.ID, "err", err)
+		return err
+	}
+
+	// If mariadb is running then exit.
+	if formation.Processes["mariadb"] > 0 {
+		logger.Info("database is running, scaling not necessary")
+		return nil
+	}
+
+	// Copy processes and increase database processes to 3.
+	processes := make(map[string]int, len(formation.Processes))
+	for k, v := range formation.Processes {
+		processes[k] = v
+	}
+	processes["mariadb"] = 3
+
+	// Otherwise begin scaling the database processes up.
+	logger.Info("watching job events", "app", app, "release_id", release.ID)
+	expected := client.ExpectedScalingEvents(formation.Processes, processes, release.Processes, 1)
```

Comment:
Can we just wait for the discoverd event and then poll the status endpoint for read/write?